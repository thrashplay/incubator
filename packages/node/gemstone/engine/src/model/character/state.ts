export type CharacterId = string
export interface Character {
  id: CharacterId
  name: string

  /** the actor's base speed, in feet of movement per minute */
  speed: number
}

export interface CharacterState {
  pcs: Record<CharacterId, Character>
}

export interface CharacterStateContainer {
  characters: CharacterState
}
