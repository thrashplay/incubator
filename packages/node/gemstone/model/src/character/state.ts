import { Dictionary } from '../common'

export type CharacterId = string
export interface Character {
  id: CharacterId
  name: string

  /** the actor's base reach, in feet; defaults to the rule system's default melee range */
  reach?: number

  /**
   * The actor's base size, which is the radius of the circle they occupy (other base shapes to come later).
   * Currently defaults to 3 feet.
   **/
  size: number

  /** the actor's base speed, in feet of movement per minute */
  speed: number
}

export type CharacterRecordSet = Dictionary<CharacterId, Character>

export interface CharacterStateContainer extends Dictionary<string, unknown> {
  characters: CharacterRecordSet
}
