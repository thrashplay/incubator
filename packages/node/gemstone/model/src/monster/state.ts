import { Dictionary } from '../types'

export type MonsterTypeId = string
export interface MonsterType {
  id: MonsterTypeId
  name: string

  /** the monster's base reach, in feet; defaults to the rule system's default melee range */
  reach?: number

  /**
   * The monster's base size, which is the radius of the circle they occupy (other base shapes to come later).
   * Currently defaults to 3 feet.
   **/
  size?: number

  /** the monster's base speed, in feet of movement per minute */
  speed: number
}

export interface MonsterState {
  types: Dictionary<MonsterTypeId, MonsterType>
}

export interface MonsterStateContainer {
  monsters: MonsterState
}
