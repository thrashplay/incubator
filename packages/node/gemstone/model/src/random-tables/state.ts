import { Dictionary } from '../common'

export type TableRollerFunction<TResult extends unknown = any> = () => TResult

export interface RandomTable<TResult extends unknown = any> {
  /** unique ID for this table */
  id: string

  /** Dice expression used to roll on this table */
  diceExpression: string

  /**
   * Default result if one does not exist, if -- for example -- the results array
   * is not exhaustive for the dice expression
   **/
  defaultResult: TResult

  /** Human-readable name of this table */
  name: string

  /** Lookup table mapping roll values to the corresponding result */
  results: Dictionary<number, TResult>
}

export type RandomTableSet = Dictionary<string, RandomTable>

export interface RandomTablesStateContainer {
  tables: RandomTableSet
}
