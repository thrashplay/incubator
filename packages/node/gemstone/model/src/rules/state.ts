import { Dictionary } from '../common'

export type MovementModeId = string

export interface MovementMode {
  id: MovementModeId
  name: string
  multiplier: number
}

export interface RuleSet {
  /** distance, in feet, that is considered "in melee" */
  meleeRange: number
  movement: {
    defaultMode?: MovementModeId
    modes: Dictionary<MovementModeId, MovementMode>
  }
  segmentDuration: number
}

export interface RulesStateContainer {
  rules: RuleSet
}
