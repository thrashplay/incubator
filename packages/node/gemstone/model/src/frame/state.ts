import { Point } from 'gemstone/model/src/common/types'

import { Character, CharacterId, CharacterStateContainer } from '../character'
import { RulesStateContainer } from '../rules'

export const EMPTY_FRAME: Frame = {
  actors: {},
  timeOffset: 0,
}

export interface IntentionType<TType extends string = string, TData extends any = any> {
  type: TType
  data?: TData
}

/** scene-specific status for the actor */
export interface ActorStatus {
  id: CharacterId

  /** the actor's current intention (that is, selected game action) */
  intention: IntentionType

  /** the current position of the actor, in world coordinates */
  position: Point
}

/** combined data structure, merging full character info with the associated scene status */
export type Actor = Character & {
  status: ActorStatus
}

/** a single frame, containing the status of all actors at a single point in a time */
export interface Frame {
  actors: Record<CharacterId, ActorStatus>

  /** the time offset, in seconds, of this frame */
  timeOffset: number
}

export interface SimulationState {
  currentFrame: Frame
}

export type ExternalRequiredState = RulesStateContainer & CharacterStateContainer
