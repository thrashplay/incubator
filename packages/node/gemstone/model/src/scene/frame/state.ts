import { Point } from 'gemstone/model/src/common/types'

import { Character, CharacterId, CharacterStateContainer } from '../../character'
import { MovementModeId, RulesStateContainer } from '../../rules'

export const EMPTY_FRAME: Frame = {
  actors: {},
  keyFrame: false,
  timeOffset: 0,
}

export const IDLE_ACTION: ActionType = { type: 'idle' }

export interface ActionType<TType extends string = string, TData extends any = any> {
  type: TType
  data?: TData
}

/** scene-specific status for the actor */
export type ActorStatus = {
  id: CharacterId

  /** the actor's current action (that is, selected game action) */
  action: ActionType

  /** the type of movement currently used by this actor, or the rule system default if undefined */
  movementMode?: MovementModeId

  /** the current position of the actor, in world coordinates */
  position: Point

  /** the actor's current target */
  target?: CharacterId
}

/** combined data structure, merging full character info with the associated scene status */
export type Actor = Character & {
  status: Omit<ActorStatus, 'id'>
}

/** a single frame, containing the status of all actors at a single point in a time */
export type Frame = {
  actors: Record<CharacterId, ActorStatus>

  /** if true, this frame is a 'key frame', meaning there is a reason to stop the simulation when it is reached */
  keyFrame?: boolean

  /** the time offset, in seconds, of this frame */
  timeOffset: number
}

export interface SimulationState {
  currentFrame: Frame
}

export type ExternalRequiredState = RulesStateContainer & CharacterStateContainer
