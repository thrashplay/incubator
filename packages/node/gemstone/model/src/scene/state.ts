import { CharacterId, CharacterStateContainer } from '../character'
import { RulesStateContainer } from '../rules'

import { Frame } from './frame'

export interface SceneState {
  /** IDs of the characters in this scene */
  characters: CharacterId[]

  /** the index of the current frame, representing the present moment */
  currentFrame: number

  /** array of frames comprising this scene */
  frames: Frame[]
}

type ExternalRequiredState = RulesStateContainer & CharacterStateContainer
export type SceneStateContainer = ExternalRequiredState & {
  scene: SceneState
}
