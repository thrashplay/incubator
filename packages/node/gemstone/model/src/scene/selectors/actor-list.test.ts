import { keys, omit } from 'lodash/fp'

import {
  ActionFixtures,
  CharacterStateFixtures,
  createStateWithDependencies,
  RulesStateFixtures,
  SceneStateFixtures,
} from '../__fixtures__'
import { EMPTY_SCENE, SceneStateContainer } from '../state'

import {
  getActor,
  getActorCollection,
  getActors,
  getActorStatusCollection,
  getCharacterIds,
} from '.'

const { GimliAndTrogdor, GimliOnly } = CharacterStateFixtures
const { Minimal } = RulesStateFixtures
const { IdleBeforeTypicalActions } = SceneStateFixtures

const defaultState: SceneStateContainer = createStateWithDependencies(IdleBeforeTypicalActions)

// this is an impossible state, but can be used to test what happens if 'frames' is somehow empty
const emptyState: SceneStateContainer = createStateWithDependencies({
  ...EMPTY_SCENE,
  frames: [],
})

const invalidState = { } as unknown as SceneStateContainer

describe('scene selectors - Actor List', () => {
  describe('getCharacterIds', () => {
    it('returns empty array if scene state is missing', () => {
      const result = getCharacterIds(invalidState)
      expect(result).toStrictEqual([])
    })

    it('returns empty array if scene has no characters', () => {
      const result = getCharacterIds(omit('characters')(emptyState) as SceneStateContainer)
      expect(result).toStrictEqual([])
    })

    it('returns correct data', () => {
      const result = getCharacterIds(defaultState)
      expect(result).toHaveLength(2)
      expect(result).toContain('gimli')
      expect(result).toContain('trogdor')
    })

    it('does not return monster IDs', () => {
      expect('it is implemented').toBe(true)
    })
  })

  describe('getActorStatusCollection', () => {
    // more cases are covered by other tests, this is just to provide visibility into if this
    // subselector is the cause of any future bugs
    it('has expected character IDs', () => {
      const result = getActorStatusCollection(defaultState)
      expect(keys(result)).toContain('gimli')
      expect(keys(result)).toContain('trogdor')
    })

    it('matches snapshot', () => {
      const result = getActorStatusCollection(defaultState)
      expect(result).toMatchSnapshot()
    })
  })

  describe('getActorCollection', () => {
    // more cases are covered by other tests, this is just to provide visibility into if this
    // subselector is the cause of any future bugs
    it('has expected character IDs', () => {
      const result = getActorCollection(defaultState)
      expect(keys(result)).toContain('gimli')
      expect(keys(result)).toContain('trogdor')
    })

    it('matches snapshot', () => {
      const result = getActorCollection(defaultState)
      expect(result).toMatchSnapshot()
    })
  })

  describe('getActors', () => {
    const stateWithoutTrogdor = ({
      characters: GimliOnly,
      rules: Minimal,
      scene: IdleBeforeTypicalActions,
    })

    const gimli = {
      id: 'gimli',
      name: 'Gimli, son of Glóin',
      speed: 60,
      status: {
        action: ActionFixtures.BefriendingElves,
        position: { x: 100, y: 100 },
      },
    }

    const trogdor = {
      id: 'trogdor',
      name: 'Trogdor, the Burninator',
      speed: 120,
      status: {
        action: ActionFixtures.Burninating,
        position: { x: 7, y: 7 },
      },
    }

    it('returns undefined if scene state is invalid', () => {
      const result = getActors(invalidState)
      expect(result).toStrictEqual([])
    })

    it('omits actors if character state is missing', () => {
      const result = getActors(stateWithoutTrogdor)
      expect(result).toHaveLength(1)
      expect(result).toContainEqual(gimli)
    })

    it('returns actors when all data is available', () => {
      const result = getActors(defaultState)
      expect(result).toHaveLength(2)
      expect(result).toContainEqual(gimli)
      expect(result).toContainEqual(trogdor)
    })
  })

  describe('getActor', () => {
    const stateWithoutTrogdor = ({
      characters: GimliOnly,
      rules: Minimal,
      scene: IdleBeforeTypicalActions,
    })

    const gimli = {
      id: 'gimli',
      name: 'Gimli, son of Glóin',
      speed: 60,
      status: {
        action: ActionFixtures.BefriendingElves,
        position: { x: 100, y: 100 },
      },
    }

    it('returns undefined if character ID is invalid', () => {
      const result = getActor(invalidState, { characterId: 'invalid-id' })
      expect(result).toBeUndefined()
    })

    it('returns undefined when character data missing', () => {
      const result = getActor(stateWithoutTrogdor, { characterId: 'trogdor' })
      expect(result).toBeUndefined()
    })

    describe('status data missing', () => {
      const sceneStateMissing = ({
        characters: GimliAndTrogdor,
        rules: Minimal,
      }) as unknown as SceneStateContainer

      it('returns undefined if scene state is missing', () => {
        const result = getActor(sceneStateMissing, { characterId: 'gimli' })
        expect(result).toBeUndefined()
      })

      it('returns undefined details if scene has no frames', () => {
        const result = getActor(emptyState, { characterId: 'gimli' })
        expect(result).toBeUndefined()
      })
    })

    describe('all data available', () => {
      it('returns correct data', () => {
        const result = getActor(defaultState, { characterId: 'gimli' })
        expect(result).toStrictEqual(gimli)
      })
    })
  })
})
