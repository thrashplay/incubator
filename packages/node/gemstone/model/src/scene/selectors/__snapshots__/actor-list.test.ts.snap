// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`scene selectors - Actor List getActorCollection matches snapshot 1`] = `
Object {
  "gimli": Object {
    "id": "gimli",
    "name": "Gimli, son of Gl√≥in",
    "speed": 60,
    "status": Object {
      "intention": Object {
        "data": "Legolas Greenleaf",
        "type": "befriending-elves",
      },
      "position": Object {
        "x": 100,
        "y": 100,
      },
    },
  },
  "trogdor": Object {
    "id": "trogdor",
    "name": "Trogdor, the Burninator",
    "speed": 120,
    "status": Object {
      "intention": Object {
        "data": Object {
          "location": "The Countryside",
          "target": "The Peasants",
        },
        "type": "burninating",
      },
      "position": Object {
        "x": 7,
        "y": 7,
      },
    },
  },
}
`;

exports[`scene selectors - Actor List getActorStatusCollection matches snapshot 1`] = `
Object {
  "gimli": Object {
    "id": "gimli",
    "intention": Object {
      "data": "Legolas Greenleaf",
      "type": "befriending-elves",
    },
    "position": Object {
      "x": 100,
      "y": 100,
    },
  },
  "trogdor": Object {
    "id": "trogdor",
    "intention": Object {
      "data": Object {
        "location": "The Countryside",
        "target": "The Peasants",
      },
      "type": "burninating",
    },
    "position": Object {
      "x": 7,
      "y": 7,
    },
  },
}
`;
